source/common/tracing/lightstep_tracer_impl.cc

```diff
@@ -15,102 +16,106 @@
 namespace Envoy {
 namespace Tracing {
 
-LightStepSpan::LightStepSpan(lightstep::Span& span, lightstep::Tracer& tracer)
-    : span_(span), tracer_(tracer) {}
+namespace {
+class LightStepLogger : Logger::Loggable<Logger::Id::tracing> {
+public:
+  void operator()(lightstep::LogLevel level, opentracing::string_view message) const {
+    const fmt::StringRef fmt_message{message.data(), message.size()};
+    switch (level) {
+    case lightstep::LogLevel::debug:
+      ENVOY_LOG(debug, "{}", fmt_message);
+      break;
+    case lightstep::LogLevel::info:
+      ENVOY_LOG(info, "{}", fmt_message);
+      break;
+    default:
+      ENVOY_LOG(warn, "{}", fmt_message);
+      break;
+    }
+  }
+};
+} // namespace
+
+LightStepDriver::LightStepTransporter::LightStepTransporter(LightStepDriver& driver)
+    : driver_(driver) {}
+
+void LightStepDriver::LightStepTransporter::Send(const Protobuf::Message& request,
+                                                 Protobuf::Message& response,
+                                                 lightstep::AsyncTransporter::Callback& callback) {
+  active_callback_ = &callback;
+  active_response_ = &response;
+
+  Http::MessagePtr message =
+      Grpc::Common::prepareHeaders(driver_.cluster()->name(), lightstep::CollectorServiceFullName(),
+                                   lightstep::CollectorMethodName());
+  message->body() = Grpc::Common::serializeBody(request);
+
+  const uint64_t timeout =
+      driver_.runtime().snapshot().getInteger("tracing.lightstep.request_timeout", 5000U);
+  driver_.clusterManager()
+      .httpAsyncClientForCluster(driver_.cluster()->name())
+      .send(std::move(message), *this, std::chrono::milliseconds(timeout));
+}
 
-void LightStepSpan::finishSpan() { span_.Finish(); }
+void LightStepDriver::LightStepTransporter::onSuccess(Http::MessagePtr&& response) {
+  try {
+    Grpc::Common::validateResponse(*response);
```

Comment:
Envoy has a gRPC client layered above `Http::AsyncClient` in https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h. Can you make use of this? This would avoid having to hand parse out gRPC messages..

Reply:
`Grpc::AsyncClientImpl` can't be used with an abstract base class for the response since it [constructs](https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h#L142) an instance of the class.

I'd rather not expose LightStep's protobuf message types directly since that would provide less encapsulation and potentially break code using the library if they were to change.

Reply:
I discussed this a while ago with @htuch and although I think we could fixup the gRPC async client to work OK with raw messages, don't worry about this for now. I would potentially just add a TODO to look into cleaning this up in the future.

Reply:
Added TODO note.